File: api/src/test/java/com/movie/api/controller/MovieControllerTest.java
package com.movie.api.controller;

import com.movie.api.service.RateLimitService;
import com.movie.api.support.IntegrationTest;
import com.movie.common.exception.RateLimitExceededException;
import com.movie.domain.entity.Movie;
import com.movie.domain.fixture.TestFixture;
import com.movie.domain.service.MovieService;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

class MovieControllerTest extends IntegrationTest {

    @MockBean
    private MovieService movieService;

    @MockBean
    private RateLimitService rateLimitService;

    @Test
    void getCurrentMovies_Success() throws Exception {
        Movie movie = TestFixture.createMovie();
        when(movieService.getCurrentMovies()).thenReturn(List.of(movie));

        mockMvc.perform(get("/api/v1/movies/current")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("OK"))
                .andExpect(jsonPath("$.code").value("SUCCESS"))
                .andExpect(jsonPath("$.data[0].title").value(movie.getTitle()));
    }

    @Test
    void getCurrentMovies_RateLimitExceeded() throws Exception {
        doThrow(new RateLimitExceededException("Too many requests"))
                .when(rateLimitService).checkIpRateLimit(any());

        mockMvc.perform(get("/api/v1/movies/current")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isTooManyRequests())
                .andExpect(jsonPath("$.status").value("TOO_MANY_REQUESTS"))
                .andExpect(jsonPath("$.code").value("RATE_LIMIT_EXCEEDED"))
                .andExpect(jsonPath("$.message").value("Too many requests"));
    }

    @Test
    void getUpcomingMovies_Success() throws Exception {
        Movie movie = TestFixture.createMovie();
        when(movieService.getUpcomingMovies()).thenReturn(List.of(movie));

        mockMvc.perform(get("/api/v1/movies/upcoming")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("OK"))
                .andExpect(jsonPath("$.code").value("SUCCESS"))
                .andExpect(jsonPath("$.data[0].title").value(movie.getTitle()));
    }

    @Test
    void getUpcomingMovies_RateLimitExceeded() throws Exception {
        doThrow(new RateLimitExceededException("Too many requests"))
                .when(rateLimitService).checkIpRateLimit(any());

        mockMvc.perform(get("/api/v1/movies/upcoming")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isTooManyRequests())
                .andExpect(jsonPath("$.status").value("TOO_MANY_REQUESTS"))
                .andExpect(jsonPath("$.code").value("RATE_LIMIT_EXCEEDED"))
                .andExpect(jsonPath("$.message").value("Too many requests"));
    }
} 

File: movie/api/src/test/kotlin/com/example/movie/controller/MovieControllerTest.kt
package com.example.movie.controller

import com.example.movie.domain.genre.model.Genre
import com.example.movie.domain.movie.model.Movie
import com.example.movie.domain.movie.model.Rating
import com.example.movie.domain.screening.model.Screening
import com.example.movie.domain.screening.model.ScreeningStatus
import com.example.movie.domain.theater.model.Theater
import com.example.movie.service.MovieSearchUseCase
import org.junit.jupiter.api.Test
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDate
import java.time.LocalDateTime

@WebMvcTest(MovieController::class)
class MovieControllerTest {
    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockitoBean
    private lateinit var movieSearchUseCase: MovieSearchUseCase

    @Test
    fun `현재 상영중인 영화 목록을 응답한다`() {
        val currentTime = LocalDateTime.of(2024, 1, 2, 0, 0, 0)

        val theater = createTheater(currentTime)
        val screenings = listOf(
            createScreening(currentTime, currentTime.plusHours(12), theater),
            createScreening(currentTime, currentTime.plusHours(14), theater),
        )
        val movie = createMovie(currentTime, screenings)

        whenever(movieSearchUseCase.getMoviesByStatusAndTitleAndGenre(ScreeningStatus.SCHEDULED, null, null))
            .thenReturn(listOf(movie))

        mockMvc.perform(get("/api/v1/movies").param("status", "SCHEDULED"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$[0].id").value(movie.id))
            .andExpect(jsonPath("$[0].screenings.length()").value(2))
    }

    private fun createMovie(currentTime: LocalDateTime, screenings: List<Screening>): Movie {
        return Movie(
            id = 1,
            title = "title1",
            rating = Rating.ALL,
            genre = createGenre(currentTime),
            releaseDate = LocalDate.of(2024,1,1),
            thumbnailUrl = "https://image.testdb/image.jpg",
            runningTime = 120,
            screenings = screenings,
            createdBy = "test",
            createdAt = currentTime,
            updatedBy = "test",
            updatedAt = currentTime
        )
    }

    private fun createGenre(currentTime: LocalDateTime) : Genre {
        return Genre(
            id = 1,
            name = "action",
            description = "thrilling action",
            createdBy = "test",
            createdAt = currentTime,
            updatedBy = "test",
            updatedAt = currentTime
        )
    }

    private fun createScreening(currentTime: LocalDateTime, screeningTime: LocalDateTime, theater: Theater): Screening {
        return Screening(
            id = 1,
            movieId = 1,
            theater = theater,
            screeningTime = screeningTime,
            screeningEndTime = screeningTime.plusMinutes(120),
            status = ScreeningStatus.SCHEDULED,
            remainingSeatCount = 25,
            createdBy = "test",
            createdAt = currentTime,
            updatedBy = "test",
            updatedAt = currentTime
        )
    }

    private fun createTheater(currentTime: LocalDateTime): Theater {
        return Theater(
            id = 1,
            name = "cgv",
            createdBy = "test",
            createdAt = currentTime,
            updatedBy = "test",
            updatedAt = currentTime
        )
    }
}